lib/elixir/src/elixir_bitstring.erl

```diff
@@ -244,12 +244,19 @@ validate_spec(signed, [])    -> {sign, signed};
 validate_spec(unsigned, [])  -> {sign, unsigned};
 validate_spec(_, _)          -> none.
 
-expand_spec_arg(Expr, S, _OriginalS, E) when is_atom(Expr); is_integer(Expr) ->
+expand_spec_arg(_, Expr, S, _OriginalS, E) when is_atom(Expr); is_integer(Expr) ->
   {Expr, S, E};
-expand_spec_arg(Expr, S, _OriginalS, #{context := match} = E) ->
-  {EExpr, SE, EE} = elixir_expand:expand(Expr, S#elixir_ex{prematch=raise}, E#{context := nil}),
+expand_spec_arg(size, Expr, S, _OriginalS, #{context := match} = E) ->
+  {EExpr, SE, EE} = elixir_expand:expand(Expr, S#elixir_ex{prematch=raise, bitsize=true}, E#{context := guard}),
   {EExpr, SE#elixir_ex{prematch=S#elixir_ex.prematch}, EE#{context := match}};
-expand_spec_arg(Expr, S, OriginalS, E) ->
+expand_spec_arg(size, Expr, S, OriginalS, E) ->
+  NewS = elixir_env:reset_read(S, OriginalS),
+  {EExpr, SE, EE} = elixir_expand:expand(Expr, NewS#elixir_ex{bitsize=true}, E#{context := guard}),
```

Comment:
Outside of a match, size can be anything. For example, this works in Erlang:

    <<255:(hd(lists:flatten([3])))>>.

So you can remove this clause and both `size` and `unit` can be handled the same outside matches.

Reply:
Let's make sure to add a test to this case. :)

Reply:
I need to double-check here, before proceeding to also add support to guard expressions inside the `unit` specifier.

At a first glace, looking at the code, I found that the `unit` case was much more restrictive than the `size`. See this example with current Elixir:

```ex
iex(1)> add = 3
iex(2)> <<ff::size(add)-unit(8)>> = <<1,2,3>>
<<1, 2, 3>>
iex(3)> <<ff::size(8)-unit(add)>> = <<1,2,3>>
** (CompileError) iex:3: unit in bitstring expects an integer as argument, got: add
```

See here: https://github.com/elixir-lang/elixir/blob/main/lib/elixir/src/elixir_bitstring.erl#L271-L272, where we are restricting `unit` expressions to be just an integer literal, I think.

So, I decided to scope this PR to work with `size` rather than `unit`. I think it is much more realistic to use more elaborate code to calculate the `size`, because the unit tends to be a fixed integer (it's an assumption, hard to really `now`).

Not fully strong to extend support to the `unit` specifier, happy to help with that in this PR or a separate one, but wanted to share my current understanding of the implications. Moreover, we would need also to change here https://github.com/elixir-lang/elixir/blob/main/lib/elixir/src/elixir_erl_pass.erl#L547-L548, right?

Reply:
Forgot to mention that I restored the original behavior in `expand_spec_arg` for the `unit` case to avoid raising errors related to invalid guard expressions when beforehand the code will flag any `unit` expression that is not a literal integer.. if we want to support more expressions for `unit`, we have to make it completely 😄

Reply:
AFAIK the behaviour is only about `size` indeed. How `unit` works does not change!